crates/lakekeeper/src/service/authz/implementations/openfga/migration.rs

```diff
@@ -2,41 +2,84 @@ use std::sync::LazyLock;
-    migration::{AuthorizationModelVersion, MigrationFn},
+    migration::{AuthorizationModelVersion, MigrationFn, TupleModelManager},
 };
 
 use super::{OpenFGAError, OpenFGAResult, AUTH_CONFIG};
+use crate::service::ServerId;
 
 pub(super) static ACTIVE_MODEL_VERSION: LazyLock<AuthorizationModelVersion> =
-    LazyLock::new(|| AuthorizationModelVersion::new(3, 4)); // <- Change this for every change in the model
+    LazyLock::new(|| *V4_MODEL_VERSION); // <- Change this for every change in the model
+
+pub(super) static V4_MODEL_VERSION: LazyLock<AuthorizationModelVersion> =
+    LazyLock::new(|| AuthorizationModelVersion::new(4, 0));
+
+#[cfg(test)]
+pub(super) static V3_MODEL_VERSION: LazyLock<AuthorizationModelVersion> =
+    LazyLock::new(|| AuthorizationModelVersion::new(3, 4));
+
+mod migration_fns_v4;
+use migration_fns_v4::{v4_push_down_warehouse_id, MigrationState};
 
 fn get_model_manager(
     client: &BasicOpenFgaServiceClient,
     store_name: Option<String>,
-) -> openfga_client::migration::TupleModelManager<BasicAuthLayer> {
-    openfga_client::migration::TupleModelManager::new(
+) -> TupleModelManager<BasicAuthLayer, MigrationState> {
+    let manager = TupleModelManager::new(
         client.clone(),
         &store_name.unwrap_or(AUTH_CONFIG.store_name.clone()),
         &AUTH_CONFIG.authorization_model_prefix,
-    )
-    .add_model(
+    );
+    // Assume vX.Y has a migration function. Then vX.Y must remain here as long as migrations from
+    // versions < vX.Y are supported.
+    add_model_v4(manager)
+}
+
+/// Has no migration hooks.
+#[cfg(test)]
+pub(crate) fn add_model_v3(
+    manager: TupleModelManager<BasicAuthLayer, MigrationState>,
+) -> TupleModelManager<BasicAuthLayer, MigrationState> {
+    manager.add_model(
         serde_json::from_str(include_str!(
             // Change this for backward compatible changes.
             // For non-backward compatible changes that require tuple migrations, add another `add_model` call.
             "../../../../../../../authz/openfga/v3.4/schema.json"
         ))
         // Change also the model version in this string:
         .expect("Model v3.4 is a valid AuthorizationModel in JSON format."),
-        AuthorizationModelVersion::new(3, 4),
+        *V3_MODEL_VERSION,
+        // For major version upgrades, this is where tuple migrations go.
+        None::<MigrationFn<_, _>>,
+        None::<MigrationFn<_, _>>,
+    )
+}
+
+/// Does have a migration hook which may add tuples to the store.
+pub(crate) fn add_model_v4(
+    manager: TupleModelManager<BasicAuthLayer, MigrationState>,
+) -> TupleModelManager<BasicAuthLayer, MigrationState> {
+    manager.add_model(
+        serde_json::from_str(include_str!(
+            // Change this for backward compatible changes.
+            // For non-backward compatible changes that require tuple migrations, add another `add_model` call.
+            "../../../../../../../authz/openfga/v4.0/schema.json"
```

Comment:
Should we fail if the `V4_MODEL_VERSION` is not part of the path?
Has always been an easy thing to miss, and even more so now with this beeing in a separate function.